runtime/gc_vlhgc/VLHGCAccessBarrier.cpp

```diff
@@ -247,45 +250,74 @@ MM_VLHGCAccessBarrier::postStoreClassToClassLoader(J9VMThread *vmThread, J9Class
 	VM_VMAccess::inlineEnterVMFromJNI(vmThread);
+	GC_ArrayObjectModel *indexableObjectModel = &_extensions->indexableObjectModel;
 
 	J9IndexableObject *arrayObject = (J9IndexableObject*)J9_JNI_UNWRAP_REFERENCE(array);
-	bool shouldCopy = false;
 	bool alwaysCopyInCritical = (javaVM->runtimeFlags & J9_RUNTIME_ALWAYS_COPY_JNI_CRITICAL) == J9_RUNTIME_ALWAYS_COPY_JNI_CRITICAL;
 	if (alwaysCopyInCritical) {
-		shouldCopy = true;
-	} else if (!_extensions->indexableObjectModel.isInlineContiguousArraylet(arrayObject)) {
+		copyArrayCritical(vmThread, indexableObjectModel, functions, &data, arrayObject, isCopy);
+	} else if (!indexableObjectModel->isInlineContiguousArraylet(arrayObject)) {
 		/* an array having discontiguous extents is another reason to force the critical section to be a copy */
-		shouldCopy = true;
-	}
-
-	if (shouldCopy) {
-		GC_ArrayObjectModel* indexableObjectModel = &_extensions->indexableObjectModel;
-		I_32 sizeInElements = (I_32)indexableObjectModel->getSizeInElements(arrayObject);
-		UDATA sizeInBytes = indexableObjectModel->getDataSizeInBytes(arrayObject);
-		data = functions->jniArrayAllocateMemoryFromThread(vmThread, sizeInBytes);
-		if(NULL == data) {
-			functions->setNativeOutOfMemoryError(vmThread, 0, 0);	// better error message here?
-		} else {
-			indexableObjectModel->memcpyFromArray(data, arrayObject, 0, sizeInElements);
-			if(NULL != isCopy) {
-				*isCopy = JNI_TRUE;
+#if defined(J9VM_GC_ENABLE_DOUBLE_MAP)
+		MM_EnvironmentVLHGC *env = MM_EnvironmentVLHGC::getEnvironment(vmThread);
+		if (indexableObjectModel->isDoubleMappingEnabled()) {
+			if (indexableObjectModel->numArraylets(arrayObject) > 1) {
+				GC_SlotObject objectSlot(env->getOmrVM(), &indexableObjectModel->getArrayoidPointer(arrayObject)[0]);
+				J9Object *firstLeafSlot = objectSlot.readReferenceFromSlot();
+				MM_HeapRegionDescriptorVLHGC *firstLeafRegionDescriptor = (MM_HeapRegionDescriptorVLHGC *)_extensions->heapRegionManager->tableDescriptorForAddress(firstLeafSlot);
+				data = firstLeafRegionDescriptor->_identifier.address;
+
+				if (NULL == data) {
+					/* Doublemap failed, but we still need to continue execution; therefore fallback to previous approach */
+					copyArrayCritical(vmThread, indexableObjectModel, functions, &data, arrayObject, isCopy);
+				}
+			/* Corner case where there's only one arraylet leaf */
+			} else if ((indexableObjectModel->numArraylets(arrayObject) == 1)
+					&& (indexableObjectModel->getSizeInElements(arrayObject) > 0)) {
+				/* Solo arraylet leaf is contiguous so we can simply return the data associated with it */
+				MM_JNICriticalRegion::enterCriticalRegion(vmThread, true);
+				Assert_MM_true(vmThread->publicFlags & J9_PUBLIC_FLAGS_VM_ACCESS);
+				GC_SlotObject objectSlot(env->getOmrVM(), &indexableObjectModel->getArrayoidPointer(arrayObject)[0]);
+				data = objectSlot.readReferenceFromSlot();
 			}
+		} else
+#endif /* J9VM_GC_ENABLE_DOUBLE_MAP */
+		{
```

Comment:
this curly open bracket should be inside endif

Reply:
This is a bit confusing but this curly bracket is part of the inner if else statement in case it exists depending on `J9VM_GC_ENABLE_DOUBLE_MAP`. If it goes inside the `#endif` and `J9VM_GC_ENABLE_DOUBLE_MAP` is not defined the compiler might complain because of missing bracket

Reply:
I was wrong indeed, I thought this else is from 'if !contiguous', but in fact it's from 'if doublemapping enabled'